File: lintcode/medium/200_longest_palindromic_substring.py
'''
Link: https://www.lintcode.com/problem/longest-palindromic-substring/description

'''
# This is a O(n^3) time complexity brute-force solution.
class Solution:
    """
    @param s: input string
    @return: the longest palindromic substring
    """
    def is_palindrome(self, s, left, right):
        while left <= right:
            if s[left] != s[right]:
                return False
            left += 1
            right -= 1
        return True
        
    def longestPalindrome(self, s):
        # write your code here
        longest_palindrome = ''
        for start_ind in range(len(s)):
            if len(longest_palindrome) >= len(s) - start_ind:
                break
            for offset in range(len(s)-start_ind):
                substr = s[start_ind: (start_ind + offset + 1)]
                if self.is_palindrome(s, start_ind, start_ind + offset):
                    if len(substr) > len(longest_palindrome):
                        longest_palindrome = substr
        return longest_palindrome

'''
This solution enumerates the middle entry of each palindromic substring. It achieves a time complexity of O(n^2), because
it no longer performs is_palindrome() test on each possible length, thus removing 1 level of nested loop in the previous
solution. It is similar to a two-pointer solution.
'''  
class Solution:
    """
    @param s: input string
    @return: the longest palindromic substring
    """
    
    def longestPalindrome(self, s):
        # write your code here
        if s is None:
            return ''
        self.start = 0
        self.longest = 0
        for middle in range(len(s)):
            # Enumerate the middle entry of palindromic substring
            self.get_palindromic_string(s, middle, middle)
            # Do the same thing for a possible even-length palindromic string.
            self.get_palindromic_string(s, middle, middle + 1)
            
        return s[self.start : (self.start + self.longest)]
    
    def get_palindromic_string(self, s, left, right):
        while left >= 0 and right < len(s) and s[left] == s[right]:
            left -= 1
            right += 1
        # Because left and right are overly incremented/decremented, we need to 
        # use the length previous to the last increment/decrement
        if self.longest < right - left - 1:
            self.longest = right - left - 1
            self.start = left + 1 

File: lintcode/medium/200_longest_palindromic_substring.java
/*
These codes are based on the python implementations. The first code block is using enumeration of central entries, which has
time complexity of O(n^2). The second code block is using Dynamic Programming, also of complexity O(n^2).
*/

public class Solution {
    /**
     * @param s: input string
     * @return: the longest palindromic substring
     */
    private int start;
    private int longest;
    public String longestPalindrome(String s) {
        // write your code here
        if (s == null)
            return "";
        start = 0;
        longest = 0;
        for (int mid = 0; mid < s.length(); mid++){
            getPalindromeString(s, mid, mid);
            getPalindromeString(s, mid, mid + 1);
        }
        return s.substring(start, start + longest);
    }
    
    private void getPalindromeString(String s, int left, int right){
        while (left >= 0 && right < s.length() && s.charAt(left) == s.charAt(right)){
            left--;
            right++;
        }
        if (longest < right - left - 1){
            longest = right - left - 1;
            start = left + 1;
        }
    } 
}

// Using dynamic programming.
public class Solution {
    /**
     * @param s: input string
     * @return: the longest palindromic substring
     */
   
    public String longestPalindrome(String s) {
        // write your code here
        if (s == null || s.length() == 0)
            return "";
        int start = 0;
        int longest = 0;
        int n = s.length();
        boolean[][] is_palindrome_tbl = new boolean[n][n];
        for (int i = 0; i < n; i++){
            is_palindrome_tbl[i][i] = true;
            
            if (i < 1){
                continue;
            }
            is_palindrome_tbl[i][i-1] = true;
        }
        longest = 1;
        for (int i = n - 1; i >= 0; i--){
            for (int j = i + 1; j < n; j++){
                if (s.charAt(i) == s.charAt(j) && is_palindrome_tbl[i+1][j-1] == true){
                    is_palindrome_tbl[i][j] = true;
                }
                if (is_palindrome_tbl[i][j] == true && j - i + 1 > longest){
                    longest = j - i + 1;
                    start = i;
                }
            }
        }
        
        return s.substring(start, start + longest);
    }    
}
